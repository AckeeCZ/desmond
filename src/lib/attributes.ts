import lodashIsempty from 'lodash.isempty';
import { isObject } from './internal/validators';
import parseBool from './parseBool';

/**
 * Transform entity attributes, as you fetch them from your storage
 *
 * 1. Null parse JSON collumns
 * 2. Put null instead of empty strings or objects
 * 3. Parse to booleans (e.g. SMALL_INT)
 */
const attributes = (input: any, params: {
    jsonColumns?: string[];
    toBoolean?: string[];
    nullOnEmpty?: '*' | string[];
} = {}) => {
    (params.jsonColumns || []).map(column => {
        if (input[column]) {
            input[column] = Array.isArray(input[column]) ? input[column] : JSON.parse(input[column]);
        }
        if (Array.isArray(input[column])) {
            input[column] = input[column].filter((attr: any) => attr !== '');
        }
    });

    (params.toBoolean || []).map(column => {
        if (input[column] !== undefined) {
            input[column] = parseBool(input[column]);
        }
    });

    const nullOnEmptyCols = Array.isArray(params.nullOnEmpty) ? params.nullOnEmpty : Object.keys(input).filter(col => isObject(input[col]));

    nullOnEmptyCols.forEach(column => {
        if (lodashIsempty(input[column])) {
            input[column] = null;
        }
    });

    return input;
};

export default attributes;
