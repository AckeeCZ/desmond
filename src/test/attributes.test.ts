import attributes from 'lib/attributes';

const data = {
    someIds: '[1,2,3,4]',
    active: 1,
    boring: 0,
    nullThis: '',
    keepThis: [],
};

describe('attributes', () => {
    test('Is unchanged without options', () => {
        expect(attributes(data)).toEqual(data);
    });
    test('Converts JSON, bool and omits nulls', () => {
        expect(attributes(data, {
            jsonColumns: ['someIds'],
            toBoolean: ['active', 'boring'],
            nullOnEmpty: ['nullThis'],
        })).toMatchSnapshot();
    });
});
