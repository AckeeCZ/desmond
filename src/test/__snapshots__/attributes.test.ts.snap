// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`attributes Converts JSON, bool and omits nulls 1`] = `
Object {
  "active": true,
  "boring": false,
  "keepThis": Array [],
  "keepThisDate": 1970-01-01T00:00:00.000Z,
  "nullThis": null,
  "someIds": Array [
    1,
    2,
    3,
    4,
  ],
}
`;
